### DISCUSSION TITLE
How do you translate an error message for a specific field?

### DISCUSSION THREAD

**Discussion by lukasvice** [colinhacks/zod#2073] [Q&A] [answered]
I would like to give a specific field its own `ErrorMap` and translate the error messages.

I define the schema at top level in the React component file, as I also need the exported inferred schema type somewhere else.

The problem is that I have to specify the `ErrorMap` when defining the schema. At this moment I don't have the context for the internationalization yet, so I can't return the correctly translated error message.

What would be the best way to solve this?

I think it would be great if the `errorMap` could be added afterwards for a field. Something like `schema.shape.fieldName.setErrorMap(...)`. Then I could define the schema at top-level and add the `errorMap` at runtime when I have all the context.

Example:

```tsx
const schema = z.object({
  confirmCheckbox: z.literal<boolean>(true, {
    errorMap: (issue, ctx) => ({
      message:
        issue.code === z.ZodIssueCode.invalid_literal
          ? 'You need to confirm' // This should be translated but I don't have the intl context here
          : ctx.defaultError,
    }),
  }),
})

export type FormData = z.infer<typeof schema>

export function Component() {
  const intl = useIntl()

  // I can use the intl context here. Would be nice to have:
  // schema.shape.confirmCheckbox.setErrorMap(...)

  return <>...</>
}
```

Thank you for any advice!

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
function makeFormData ( errorMessage?: string ) {
    return z.object( {
        confirmCheckbox: z.literal<boolean>( true, {
            errorMap: ( issue, ctx ) => ( {
                message: issue.code === z.ZodIssueCode.invalid_literal
                    ? errorMessage ?? 'You need to confirm'
                    : ctx.defaultError,
            } ),
        } ),
    } )
}

type FormData = z.infer<ReturnType<typeof makeFormData>>
// type FormData = {
//     confirmCheckbox: boolean
// }

function Component () {
    const schema = makeFormData( 'some other message' )
    return <></>
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by lukasvice** [reply]
Oh nice, that's it! Thank you so much!
